From: Christoph M. Becker <cmbecker69@gmx.de>
Date: Sat Dec 29 14:17:23 2018 +0100
Subject: [PATCH] Fix #77367: Negative size parameter in mb_split

When adding the last element to the result value of `mb_split`, the
`chunk_pos` may point beyond the end of the string, in which case the
unsigned `n` would underflow.  Therefore, we check whether this is the
case in the first place, and only calculate `n` otherwise.  Since `n`
is no longer used outside the block, we move its declaration inside.

5.4 backport: https://github.com/shivammathur/php-src-backports/commit/27d8d19316d3e7f5cab73c37ce825345b870ae1b
---
diff --git a/ext/mbstring/php_mbregex.c b/ext/mbstring/php_mbregex.c
--- a/ext/mbstring/php_mbregex.c
+++ b/ext/mbstring/php_mbregex.c
@@ -1057,7 +1057,7 @@ PHP_FUNCTION(mb_split)
 	OnigUChar *pos, *chunk_pos;
 	int string_len;
 
-	int n, err;
+	int err;
 	long count = -1;
 
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss|l", &arg_pattern, &arg_pattern_len, &string, &string_len, &count) == FAILURE) {
@@ -1115,8 +1115,8 @@ PHP_FUNCTION(mb_split)
 	}
 
 	/* otherwise we just have one last element to add to the array */
-	n = ((OnigUChar *)(string + string_len) - chunk_pos);
-	if (n > 0) {
+	if ((OnigUChar *)(string + string_len) > chunk_pos) {
+		size_t n = ((OnigUChar *)(string + string_len) - chunk_pos);
 		add_next_index_stringl(return_value, (char *)chunk_pos, n, 1);
 	} else {
 		add_next_index_stringl(return_value, "", 0, 1);
